routing_common/transit_types.hpp

```diff
@@ -25,16 +25,17 @@ using TransferId = uint64_t;
 using Weight = double;
 using Ranges = std::vector<std::vector<StopId>>;
 
+Anchor constexpr kInvalidAnchor = std::numeric_limits<Anchor>::max();
+std::string const kInvalidColor = std::string("");
```

Comment:
Note - it's possible to use boost::optional.

Reply:
Thanks. Which advantages will we get if we use it here? Or it's just an alternative?
Will we need to have an operator(boost::optional<std::string> ...) in visitors to process boost::optional field? Please note that color is not an optional field at json. If "color" is not found we stop the generator in BuildTransit() function with a CHECK.

Reply:
It looks like you need to do following in, say, Json serializer:
```c++
template <typename T>
void Visit(optional<T> & value, char const * name)
{
  if (json has field name)
    (*this)(*value, name);
  else
    value.clear();
}
```

Reply:
Yes, thanks! it's exactly a place where it's worth using optional I guess. 
Json may contain and may not contain a field. 

In our case if a field of a type declared in routing_common/transit_types.hpp is optional according to our json, the filed should be declared as optional<T> in the class.
If not the field is used without `optional`. For example:
```c++
class Stop
{
public:
  Stop() : m_featureIdentifiers(true /* serializeFeatureIdOnly */) {};
  Stop(...);
...

private:
  DECLARE_TRANSIT_TYPE_FRIENDS
  DECLARE_VISITOR_AND_DEBUG_PRINT(...)

  StopId m_id = kInvalidStopId;
  boost::optional<FeatureIdentifiers> m_featureIdentifiers;
  boost::optional<TransferId> m_transferId;
  std::vector<LineId> m_lineIds;
  boost::optional<m2::PointD> m_point;
  boost::optional<std::vector<TitleAnchor>> m_titleAnchors;
};
```
Do you propose something like that? I let us to get rid of `kInvalid*` constants.

Reply:
Yes, it's exactly what I'm proposing. Looks like it's not too hard to implement, what do you think?

Reply:
Yes. I guess it's worth implementing it. As far as I understand it shouldn't be too hard. I like this approach because in that case we show explicitly which json fields are optional and which not.

I'll try to find resources to implement it.